serialize/types.go

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package serialize
+
+import (
+	"math"
+
+	"github.com/m3db/m3x/checked"
+	"github.com/m3db/m3x/ident"
+)
+
+var (
+	// MaxNumberTags is the maximum number of tags that can be encoded.
+	MaxNumberTags = math.MaxUint16
+
+	// MaxTagLiteralLength is the maximum length of a tag Name/Value.
+	MaxTagLiteralLength uint16 = 4096
+
+	// HeaderMagicNumber is an internal header used to denote the beginning of
+	// an encoded stream.
+	HeaderMagicNumber uint16 = 10101
+)
+
+// Encoder encodes provided Tag iterators.
+type Encoder interface {
+	// Encode encodes the provided iterator into it's internal byte stream.
+	// NB: leaves the original iterator un-modified.
+	Encode(ident.TagIterator) error
+
+	// Data returns the encoded bytes.
+	// NB: Data() requires Encode() be called first, and it returning no
+	// error. Calling Data() without condition will cause a panic.
```

Comment:
Hm, perhaps better to return an error rather than panic.  We have tried very hard to make sure, even if it makes the API awkward, to always return err instead of panic.

Reply:
Made the change.

Reply:
Makes sense, although I don't see any panics in checked bytes?

Reply:
`checked.Bytes` uses ReadWriteRef which throws panics if the ref counting isn't done correctly. e.g. https://github.com/m3db/m3x/blob/master/checked/ref.go#L98